![CC-BY-SA](https://mirrors.creativecommons.org/presskit/buttons/88x31/svg/by-sa.svg)


This notebook was created by [Bernardo Freitas Paulo da Costa](http://www.im.ufrj.br/bernardofpc),
and is licensed under Creative Commons BY-SA.

Antes de enviar este Teste, verifique que tudo está funcionando como esperado.
Por exemplo, **rode o código inteiro, do zero**.
Para isso, vá no menu, escolha _Kernel_, depois _Restart & Run All_.

Verifique, também, que você respondeu todas as questões:
* as questões de código têm `YOUR CODE HERE` (e você pode apagar o `raise NotImplemented` ao incluir sua resposta)
* as questões discursivas têm "YOUR ANSWER HERE".

---

In [1]:
import numpy as np

# Teste 1: Bisseção

## Questão 1: Número de iterações da bisseção, e de chamadas à função `f`

Generalize o algoritmo da bisseção _iterativa_ para retornar,
- a (aproximação da) raiz,
- o número de bisseções (divisões por 2 do intervalo) feitas,
- o número de vezes que você chamou a função `f`

Use como critério de parada
- `xtol`, e
- `maxiter`,

ou seja, o algoritmo termina quando:
- seja possível garantir que o erro (absoluto) da resposta ("em $x$") seja menor do que `xtol`; OU
- quando o algoritmo já tiver feito `maxiter` itererações.

In [2]:
def bissecao(f, a, b, xtol=1e-8, maxiter=10):
    """A função recebe um function <f>, um float <a> e um float <b>. Podem ser passados também o float <xtol> e o int <maxiter>.
    Função que a partir de uma função <f>, um limite inferior <a> e um limite superior <b>, calcula a raiz aproximada da função 
    usando o método de bisseção, sendo esta obtida a partir de uma tolerância <xtol> (1e-8 como padrão) ou a partir de um número
    máximo de divisões desse intervalo <maxiter> (10 como padrão).
    A função retorna um float(a raiz) e dois ints(o número de divisões do intervalo e o número de chamadas da função <f>)"""
    assert f(a)*f(b) < 0, "O intervalo foi mal escolhido"
    niters = 0
    ncalls = 2
    while abs(b-a)>=xtol:
        meio = (a+b)/2
        if f(meio)*f(a)<0:
            b = meio
        else:
            a = meio
        ncalls += 2
        niters += 1
        if niters >= maxiter:
            break
    return (a+b)/2, niters, ncalls

### 1.1: Testando com $\sqrt2$:

In [3]:
def p2(x): return x**2 - 2

In [4]:
raiz, niters, ncalls = bissecao(p2, 0, 2, maxiter=100)
assert abs(raiz - np.sqrt(2)) < 1e-8
assert 25 <= niters <= 30
assert 25 <= ncalls

In [5]:
raiz, niters, ncalls = bissecao(p2, 0, 2, xtol=1e-3)
assert abs(raiz - np.sqrt(2)) < 1e-3
assert 9 <= niters <= 11

In [6]:
raiz, niters, ncalls = bissecao(p2, 0, 2)
assert abs(raiz**2 - 2) > 1e-3
assert niters == 10

### 1.2: Tolerância, número de iterações, respostas, ...

Observe o seguinte código:

In [7]:
for err in np.logspace(-8, 0, num=9):
    r, ni, nc = bissecao(p2, 0, 2, xtol=err)
    print(f"xtol = {err:.1e} --> raiz = {r:.10f}, #iter = {ni:3d}")

xtol = 1.0e-08 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-07 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-06 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-05 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-04 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-03 --> raiz = 1.4150390625, #iter =  10
xtol = 1.0e-02 --> raiz = 1.4179687500, #iter =   8
xtol = 1.0e-01 --> raiz = 1.4062500000, #iter =   5
xtol = 1.0e+00 --> raiz = 1.2500000000, #iter =   2


Comente

Com a tolerância cada vez menor, é muito mais fácil atingir o limite de $10$ iterações. Sendo assim, com tolerâncias menores ou iguais a $10^{-3}$, o valor aproximado da raiz não consegue passar de $1.4150390625$

Como seria o equivalente se o critério fosse o número de iterações?

In [8]:
for iters in range(10,36,5):
    r, ni, nc = bissecao(p2, 0, 2, maxiter=iters)
    print(f"niters = {ni} --> raiz = {r:.10f}")

niters = 10 --> raiz = 1.4150390625
niters = 15 --> raiz = 1.4142150879
niters = 20 --> raiz = 1.4142141342
niters = 25 --> raiz = 1.4142135680
niters = 28 --> raiz = 1.4142135642
niters = 28 --> raiz = 1.4142135642


Comente, também.

A partir de $28$ iterações, o valor da tolerância fica menor que $10^{-8}$, que é o padrão. Sendo assim, o valor aproximado da raiz não passa de $1.4142135642$

### 1.3: Agora, um polinômio um pouco mais complicado:

In [9]:
def p4(x): return x**4 - 3*x + 1

In [10]:
r1, n1, c1 = bissecao(p4, 0, 1, maxiter=100)
assert p4(r1) < 1e-6

In [11]:
r2, n2, c2 = bissecao(p4, 1, 2, maxiter=100)
assert n1 == n2
assert p4(r2) < 1e-6

In [12]:
r3, n3, c2 = bissecao(p4, 1, 3, maxiter=100)
assert n3 == n2+1
assert r3 == r2

### 1.4: Relação entre `ncalls` e `niters`

Muito provavelmente, é possível calcular `ncalls` a partir de `niters`.
Qual a relação, para o seu programa, entre estes valores?
Porquê?

In [13]:
def ncalls(niters):
    """A função utiliza a relação entre ncalls e niters para calcular o valor de ncalls"""
    return 2*niters + 2

No meu código, o número de chamadas já começa em 2, devido a um assert. Dentro de cada loop do while, para cada contagem do número de iterações duas contagens são feitas para o número de chamadas. Sendo assim, o número de chamadas é o dobro do número de iterações somado às duas chamadas iniciais do assert.

## Questão 2: Bisseção em listas ordenadas

Se desejamos encontrar um número em uma lista, podemos percorrer cada um de seus índices
até achar o número, e retornamos o índice correspondente.

Mas, se a lista `l` possui números em ordem,
é possível usar uma variante da bisseção para encontrar este índice.
Observe que agora desejamos retornar um número **inteiro** (o índice!),
e que índices para listas sempre devem ser inteiros, portanto:
- cuidado ao escolher o "ponto médio"
- mas o critério de parada é "quando achar o índice".

### 2.1: Encontrando um número

Implemente a função `bissect_list(l, v)`, que retorna
- o índice `k` do elemento na lista `l`, que vale `v`; e
- o número de acessos à lista `l`.

Suponha que a lista `l` é não-decrescente (ou seja, `l[i] <= l[i+1]`).

In [14]:
def bissect_list(l, v):
    """Índice  k  na lista  l , crescente, tal que l[k] = v.  None caso não exista."""
    n = len(l)
    naccess = 0
    lista_aux = [a for a in range(n)]
    while n >= 1:
        meio = l[n//2]
        naccess += 1
        if meio == v:
            return lista_aux[n//2], naccess
        elif meio < v:
            l = l[n//2:]
            lista_aux = lista_aux[n//2:]
        else:
            l = l[:n//2]
            lista_aux = lista_aux[:n//2]
        n = len(l)
        if n == 1 and v not in l:
            break
    return None, naccess

In [15]:
assert bissect_list([], 42) == (None,0)

In [16]:
l = [4]
assert bissect_list(l, 4) == (0,1)

In [17]:
assert bissect_list(l, 5) == (None,1)

In [18]:
l = [1,3,6,10,15,21,28]
for i,v in enumerate(l):
    idx, n = bissect_list(l, v)
    assert idx == i, f"v={v}, i={i}, idx={idx}"
    print(n)

3
2
3
1
3
2
3


Qual o maior número de acessos à lista que você precisou para encontrar um valor nela?
Este número depende da posição na lista?  Explique!

O maior valor de acessos foi $3$, e este valor depende da posição na lista. Isso ocorre devido ao método da busca utilizado na função, em algum momento o valor que estou procurando estará na exata metade de uma lista. O valor $3$, por exemplo, não é a metade da primeira lista (que no caso é $10$) mas é a metade da segunda, formada por $[1,3,6]$. Sendo assim, o número de acessos depende de quão próximo de uma das metades o seu número está.

### 2.2: Uma lista muito maior

In [19]:
l = np.arange(0, 8)
for v in l:
    idx, n = bissect_list(l, v)
    assert l[idx] == v
    assert n <= 20
    print(f"v={v:5d}, idx={idx:5d}, n={n:3d}")

v=    0, idx=    0, n=  4
v=    1, idx=    1, n=  3
v=    2, idx=    2, n=  2
v=    3, idx=    3, n=  3
v=    4, idx=    4, n=  1
v=    5, idx=    5, n=  3
v=    6, idx=    6, n=  2
v=    7, idx=    7, n=  3


In [20]:
l = np.arange(-4000, 6000)
np.random.seed(1)
valores_testes = np.random.randint(low=-4000, high=6000, size=10)
for v in valores_testes:
    idx, n = bissect_list(l, v)
    assert l[idx] == v
    assert n <= 20
    print(f"v={v:5d}, idx={idx:5d}, n={n:3d}")

v=-3765, idx=  235, n= 13
v= 1192, idx= 5192, n= 12
v=-3095, idx=  905, n= 13
v= 3813, idx= 7813, n= 13
v=-1105, idx= 2895, n= 13
v= 1056, idx= 5056, n= 13
v=-3856, idx=  144, n= 13
v=  225, idx= 4225, n= 13
v= 3751, idx= 7751, n= 13
v= -538, idx= 3462, n= 13


Como o número de acessos varia, nesta outra lista?
E como ele se compara com o tamanho da lista?

O número de acessos varia de $1$ a $14$, de acordo com a posição do valor na lista. Este número tem como valor máximo $14$ devido ao tamanho da lista. Como a lista tem tamanho $10^{4}$, o número máximo é o maior $x\in\mathbb{N}$ que resolve $\frac{len(l)}{2^{x-1}}\ge1$, pois o tamanho das listas após a bisseção está atrelado à $\frac{len(l)}{2^{k}}$.

### 3.3: Listas crescentes e decrescentes.

Altere o código anterior, para receber um argumento a mais, `decr`, que indica que a lista é **decrescente**.
Não reordene a lista!

In [21]:
def bissect(l, v, decr=False):
    """Index  k  on a monotonic list  l  such that  l[k] == v.
    Returns  None  if  not present.
    If  l  is decresing, pass  decr=True."""
    n = len(l)
    naccess = 0
    lista_aux = [a for a in range(n)]
    while n >= 1:
        meio = l[n//2]
        naccess += 1
        if decr:
            if meio == v:
                return lista_aux[n//2], naccess
            elif meio > v:
                l = l[n//2:]
                lista_aux = lista_aux[n//2:]
            else:
                l = l[:n//2]
                lista_aux = lista_aux[:n//2]
        else:
            if meio == v:
                return lista_aux[n//2], naccess
            elif meio < v:
                l = l[n//2:]
                lista_aux = lista_aux[n//2:]
            else:
                l = l[:n//2]
                lista_aux = lista_aux[:n//2]
        n = len(l)
        if n == 1 and v not in l:
            break
    return None, naccess

In [22]:
l = [1,3,6,10,15,21,28,35,36]
assert bissect(l, 15)[0] == 4

In [23]:
l_rev = [36, 35, 28, 21, 15, 10, 6, 3, 1]
assert bissect(l_rev, 15, decr=True)[0] == 4

In [24]:
l_rev = [36, 35, 28, 21, 15, 10, 6, 3, 1]
assert bissect(l_rev, 16, decr=True)[0] == None

In [25]:
l3 = [36, 35, 28, 21, 15, 10, 9, 6, 3, 2, 1]
assert bissect(l3, 15, decr=True)[0] == 4

### 2.4: Bônus

O número de acessos à lista para encontrar um número muda se esta está invertida?
Explique!

Sim! Se o tamanho da lista for par, é mais rápido achar o número na última posição do que na primeira. Isso ocorre pois na hora de dividir uma lista par, o número escolhido como "metade" é o maior das duas opções. 

In [26]:
l_1 = [1,3,6,10,15,21,28,35]
l_2 = [35,28,21,15,10,6,3,1]
for c in l_1:
    a = bissect(l_1,c)[1]
    b = bissect(l_2,c,True)[1]
    print(a,b)

4 3
3 2
2 3
3 1
1 3
3 2
2 3
3 4
